### ISSUE TITLE
Designing relationships between components

### ISSUE THREAD

**Issue by Morganamilo** [ably/ably-rust#4]
I think the first thing to get sorted out before we work on how the public API should look is how the structs should look. In particular, the `Rest`, `Client` and `Auth` types.

Right now there's some weird circular logic, 

auth contains a client
client contains an auth
rest contains an auth and a client

Not only is this a bit hard to comprehend, but it also duplicates the data multiple times.

At a glance, reference counting is the obvious choice here. We can have everything point at each other without duplicating data. However this doesn't solve everything as 2 way ref counting doesn't really work too well in rust. If you have two Rcs pointing to each other then the ref count never drops to 0 and they never deallocate.

You could partially solve this with weak references. But then you would have to handle the case of one of the structs being dealloced which would not only be a hassle but add possible runtime errors.

So. We have a bunch of data that we want to live equally long and be able to access each other. Why do we want them to be two separate structs in the first place? Because the spec says so!

I purpose we simply fake the spec! Say if our auth struct looks like this:

```rust
pub struct Auth<'a> {
    rest: &'a Rest
}

impl Rest {
    pub fn auth(&self) -> Auth {
        Auth { rest: self }
    }
}
```

With this method, all out data comfortably sits in one place, the auth struct pretty much just works as namespacing for our public API.

**Comment by Morganamilo** [comment]
@lmars this is quite a ramble but I just wanted to make my thought process known and have some sort of tracking of what I'm currently doing. Do let me know what you think of this strategy.

**Comment by lmars** [comment]
@Morganamilo this is all spot on, I decided to stick with the IDL in the spec which lead to this awkward design, and it really is awkward!

I think we have a few options here:

1. Break away from the IDL in ably-rust, make `auth` a method call rather than a field
2. Break away from the IDL in ably-rust, move the auth related logic up into the client and do away with the separate auth indirection
3. Update the IDL to convert things like `auth` to method calls rather than fields (which is also desirable for [mocking in ably-go](https://github.com/ably/ably-go/issues/498)), then adopt in ably-rust
4. Do away with the IDL, let each SDK implement its own idiomatic design

I'd be ok with any of these approaches at this early stage, but I think we should try and get consensus on where we're heading with the spec, the IDL, and the various SDK implementations of it.

**Comment by Morganamilo** [comment]
I'll see if @QuintinWillison thinks ignoring the IDL is too radical.

**Comment by QuintinWillison** [comment]
Ignoring [the IDL](https://docs.ably.com/client-lib-development-guide/features/#idl) is definitely not too radical, and is probably preferable. Our desire and strong preference is to meet developers where they are (i.e. for our SDK to fit into _their_ ecosystem, in this case Rust), however we also need to ensure that those reading our [primordial conceptual documentation](https://ably.com/docs/) are not going to be radically confused as to 'where to find things' in the Rust SDK... so, that's the challenge/fence you're walking. 😉

**Comment by Morganamilo** [comment]
Yes exactly. That's why I'm suggesting the auth object basically be a noop but serve as API name spacing to match how ably libs are generally structured.
